### PR TITLE
Update 5.2 to follow BASHER

### PR DESCRIPTION
**Pull request by derisen**
## Purpose

* Bump msal-angular and microsoft.identity.web
* Use single app registration for both spa and web api
* new angular groups guard for protecting components
* code refactoring and documentation improvements

## Does this introduce a breaking change?

```
[ x ] Yes
[ ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ x] Feature
[ x] Code style update (formatting, local variables)
[ x] Refactoring (no functional changes, no api changes)
[ x] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

## What to Check
Verify that the following are valid
* Overage scenario successfully handled

### CODE REVIEW

**Comment by aremo-ms** [comment]
Please add a name for logged-in user
image

File: 5-AccessControl/2-call-api-groups/README.md [hunk 1]
```diff
@@ -85,161 +105,182 @@ For more information and potential issues, see: [HTTPS in .NET Core](https://doc
...
+   1. Ensure that the **My APIs** tab is selected.
+   1. In the list of APIs, select the API `msal-angular-app`.
+        1. Since this app signs-in users, we will now proceed to select **delegated permissions**, which is is requested by apps when signing-in users.
+            1. In the **Delegated permissions** section, select the **Access 'msal-angular-app'** in the list. Use the search box if necessary.
+   1. Select the **Add permissions** button at the bottom.
+   1. Select the **Add a permission** button and then:
+   1. Ensure that the **Microsoft APIs** tab is selected.
+   1. In the *Commonly used Microsoft APIs* section, select **Microsoft Graph**
+      * Since this app signs-in users, we will now proceed to select **delegated permissions**, which is is requested by apps when signing-in users.
+           1. In the **Delegated permissions** section, select the **User.Read**, **GroupMember.Read.All** in the list. Use the search box if necessary.
```
Comment (kalyankrishna1) [inline 5-AccessControl/2-call-api-groups/README.md:203]: >**GroupMember.Read.All** [](http://example.com/codeflow?start=82&length=24)

We need a way to place the text asking for admin consent here, maybe a new non-mandatory flag named RequiresAdminConsent on the app object itself?

Comment (derisen) [reply 5-AccessControl/2-call-api-groups/README.md:203]: Good idea. Let me add it in the codegen

Comment (derisen) [reply 5-AccessControl/2-call-api-groups/README.md:203]: https://github.com/Azure-Samples/GenerateScriptsForSample/issues/44

File: 5-AccessControl/2-call-api-groups/API/TodoListAPI/Startup.cs [hunk 2]
```diff
@@ -3,65 +3,96 @@
...
+                           // calls method to process groups overage claim.
+                           await GraphHelper.GetSignedInUsersGroups(context);
+
+                           await Task.CompletedTask;
+                        };
+                    }, options => { Configuration.Bind("AzureAd", options); })
                         .EnableTokenAcquisitionToCallDownstreamApi(options => Configuration.Bind("AzureAd", options))
-                        .AddMicrosoftGraph(Configuration.GetSection("GraphBeta"))
-                    .AddInMemoryTokenCaches();
+                        .AddMicrosoftGraph(Configuration.GetSection("MsGraph"))
```
Comment (kalyankrishna1) [inline 5-AccessControl/2-call-api-groups/API/TodoListAPI/Startup.cs:72]: >MsGraph [](http://example.com/codeflow?start=69&length=7)

nit:MSGraph

File: 5-AccessControl/2-call-api-groups/API/TodoListAPI/Startup.cs [hunk 3]
```diff
@@ -3,65 +3,96 @@
...
+                // The claim in the Jwt token where Security Groups are available.
                 options.TokenValidationParameters.RoleClaimType = "groups";
             });
 
             // Adding authorization policies that enforce authorization using Azure AD roles.
             services.AddAuthorization(options =>
             {
-                options.AddPolicy(AuthorizationPolicies.AssignmentToGroupAdminGroupRequired, policy => policy.RequireRole(Configuration["Groups:GroupAdmin"]));
-                options.AddPolicy(AuthorizationPolicies.AssignmentToGroupMemberGroupRequired, policy => policy.RequireRole(Configuration["Groups:GroupMember"]));
+                options.AddPolicy(AuthorizationPolicies.AssignmentToGroupMemberGroupRequired, policy => policy.RequireRole(Configuration["AzureAd:Groups:GroupMember"], Configuration["AzureAd:Groups:GroupAdmin"]));
```
Comment (kalyankrishna1) [inline 5-AccessControl/2-call-api-groups/API/TodoListAPI/Startup.cs:86]: >Configuration["AzureAd:Groups:GroupAdmin"] [](http://example.com/codeflow?start=168&length=42)

shouldn't the AssignmentToGroupAdminGroupRequired have both GroupMember and GroupAdmin ids assigned?

Comment (derisen) [reply 5-AccessControl/2-call-api-groups/API/TodoListAPI/Startup.cs:86]: nah that would let members to access admin only routes

**Comment by derisen** [comment]
Note: missing certificate scripts and instructions

File: 5-AccessControl/2-call-api-groups/README.md [hunk 4]
```diff
@@ -1,28 +1,47 @@
...
+- msal-angular
+- microsoft-identity-web
+name: Angular single-page application calling a protected Core web API using Security Groups to implement Role-Based Access Control
+description: Angular single-page application calling a protected web API using App Roles to implement Role-Based Access Control
+urlFragment: ms-identity-javascript-angular-tutorial
+---
+
+# Angular single-page application calling a protected Core web API using Security Groups to implement Role-Based Access Control
+
+* [Overview](#overview)
```
Comment (kalyankrishna1) [inline 5-AccessControl/2-call-api-groups/README.md:21]: >*  [](http://example.com/codeflow?start=0&length=2)

Code gen is generating deployment steps, shouldn't we just use it

**Review by kalyankrishna1 — CHANGES_REQUESTED** [review]
:clock1:

File: 5-AccessControl/2-call-api-groups/API/TodoListAPI/appsettings.json [hunk 5]
```diff
@@ -1,26 +1,26 @@
...
-  "Groups": {
-    "GroupAdmin": "Enter the objectID for GroupAdmin group copied from Azure Portal",
-    "GroupMember": "Enter the objectID for GroupMember group copied from Azure Portal"
+    "ClientSecret": "Enter the Client Secret of the 'TodoListAPI' application copied from the Azure portal",
+    "Scopes": [ "access_via_group_assignments" ],
+    "Groups": {
+      "GroupAdmin": "Enter the object ID for GroupAdmin group copied from Azure Portal",
+      "GroupMember": "Enter the object ID for GroupMember group copied from Azure Portal"
+    },
+    "MsGraph": {
```
Comment (kalyankrishna1) [inline 5-AccessControl/2-call-api-groups/API/TodoListAPI/appsettings.json:12]: >MsGraph [](http://example.com/codeflow?start=5&length=7)

Look at implementing CP1 as a Graph call is involved

Comment (derisen) [reply 5-AccessControl/2-call-api-groups/API/TodoListAPI/appsettings.json:13]: Just added. Should we look into handling claims challenge in this sample or just CP1 is fine?

Comment (kalyankrishna1) [reply 5-AccessControl/2-call-api-groups/API/TodoListAPI/appsettings.json:13]: handling has to be done as well

Comment (kalyankrishna1) [reply 5-AccessControl/2-call-api-groups/API/TodoListAPI/appsettings.json:13]: APIs cant interact, so removed it

File: 5-AccessControl/2-call-api-groups/SPA/src/app/group.guard.ts [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+                let activeAccount = this.authService.instance.getActiveAccount() as AccountWithGroupClaims;
+
+                if (!activeAccount && this.authService.instance.getAllAccounts().length > 0) {
+                    activeAccount = this.authService.instance.getAllAccounts()[0] as AccountWithGroupClaims;
+                }
+
+                // check either the ID token or the graphService to see if the user is a member of any groups
+                if (!activeAccount?.idTokenClaims?.groups && this.graphService.getUser().groupIDs.length === 0) {
+                    if (activeAccount.idTokenClaims?._claim_names && activeAccount.idTokenClaims?._claim_names.groups) {
+                        window.alert('You have too many group memberships. The application will now query Microsoft Graph to get the full list of groups that you are a member of.');
```
Comment (kalyankrishna1) [inline 5-AccessControl/2-call-api-groups/SPA/src/app/group.guard.ts:57]: >window [](http://example.com/codeflow?start=24&length=6)

Differentiate the error message between overage and absence of GroupAdmin and GroupMember memberships

File: 5-AccessControl/2-call-api-groups/SPA/src/app/overage/overage.component.ts [hunk 7]
```diff
@@ -4,68 +4,68 @@ import { GraphService } from '../graph.service';
...
-          }
+                /**
+                 * Some queries against Microsoft Graph return multiple pages of data either due to server-side paging 
+                 * or due to the use of the $top query parameter to specifically limit the page size in a request. 
+                 * When a result set spans multiple pages, Microsoft Graph returns an @odata.nextLink property in 
+                 * the response that contains a URL to the next page of results.
+                 * learn more at https://docs.microsoft.com/graph/paging
+                 */
+                if (response['@odata.nextLink']) {
+                    this.handleNextPage(response['@odata.nextLink'])
```
Comment (kalyankrishna1) [inline 5-AccessControl/2-call-api-groups/SPA/src/app/overage/overage.component.ts:34]: >. [](http://example.com/codeflow?start=56&length=1)

Use Graph sDK pls

File: 5-AccessControl/2-call-api-groups/SPA/src/app/group.guard.ts (hunk 6 above)
Comment (derisen) [reply 5-AccessControl/2-call-api-groups/SPA/src/app/group.guard.ts:57]: done

**Pull request merged**